docs/concepts/architecture/cloud-controller.md

```diff
@@ -0,0 +1,242 @@
+
+### 1. Kube Controller Manager
+
+The majority of CCM's functions are derived from KCM. As mentioned in the previous section, CCM runs the following control loops
+
+1. Node Controller
+2. Route Controller 
+3. Service Controller
+4. Persistent Volume Labels Controller
+
+#### Node Controller
+
+The Node Controller is responsible for initializing a node by obtaining information about the nodes running in the cluster from the cloud provider. The node controller performs the following functions
+
+1. Initialize a node with cloud specific zone/region labels
+2. Initialize a node with cloud specific Instance details (type, size etc.)
+3. Obtain the node's network addresses (IP addresses, hostname etc.)
+4. In case a node becomes unresponsive, then it checks the cloud to see if the node has been deleted from the cloud. It proceeds to delete the node object from Kubernetes if the node was indeed deleted from the cloud.  
+
+#### Route Controller
+
+The Route Controller is responsible for configuring routes in GCE appropriately so that containers on different nodes in the Kubernetes cluster can communicate with each other. The route controller is only applicable for GCE clusters.
+
+#### Service Controller
+
+The Service Controller is responsible for listening to service create, update, and delete events. Based on the current state of the services in Kubernetes, it configures cloud load balancers (Such as ELB, or Google LB) to reflect the state of the services in Kubernetes. Additionally, it ensures that service backends for cloud load balancers are up-to-date.
+
+#### Persistent Volume Labels Controller
+
+The Persistent Volume Labels Controller applies labels on AWS EBS, and GCE PD volumes when they are created. This removes the need for users to manually set the labels on these volumes. 
+
+These labels are essential for the scheduling of pods, as these volumes are constrained to work only within the region/zone that they are in, and therefore any pod using these volumes need to be scheduled in the same region/zone.
+
+### 2. Kubelet
+
+The Node Controller contains the cloud dependent functionality of Kubelet.
+
+### 3. Kube APIServer
+
+The Persistent Volume Labels controller moves the cloud dependent functionality of Kube APIServer to CCM. 
+
+Plugin Mechanism
+------------------
+
+The Cloud Controller Manager uses Go interfaces to allow implementations from any cloud to be plugged in. Specifically, it uses the CloudProvider Interface defined [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/cloud.go)
+
+Any vendor that wants to plugin their cloud into Kubernetes only need to satisfy that interface. The implementation of the controllers and the scaffolding will be in core, but the implementation will exec out to the cloud interfaces, so long as the interface is satisfied. 
```

Comment:
The impl of the four shared controllers highlighted above and some scaffolding along with the shared cloudprovider interface will stay in core, but cloud provider-specific implementations will be built out of core, implementing the interfaces in core.